<a href="https://colab.research.google.com/github/Chero0503/Proy_codigofacilito/blob/main/Visualizacion_de_Datos_HRMuebles.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#Visualización de Datos y Análisis de la empresa "HR Muebles"


HR Muebles es una empresa dedicada a la venta de muebles y electrodomésticos en la región suroeste del estado de Puebla, México. Se desea presentar una visualización de datos del negocio con el objetivo de maximizar las ventas en el siguiente año. Se obtuvo una base de datos de la empresa la cual vamos a analizar.

#Importación de Datos

Con las siguientes lineas de código vamos a subir un archivo .csv al notebook.

In [ ]:
from google.colab import files
files.upload()


Mandamos a traer el archivo para poder mapear el dataframe.

In [ ]:
import pandas as pd
import numpy as np

df=pd.read_csv('hrmuebles_p1.csv')
df

fecha_venta  id_venta  id_producto  \
0     24/10/2013 00:00         1           38   
1     10/08/2013 00:00         2           28   
2     24/12/2012 00:00         3           34   
3     14/10/2013 00:00         6           15   
4     15/10/2013 00:00         7           12   
...                ...       ...          ...   
7927  23/12/2022 00:00     12806           14   
7928  23/12/2022 00:00     12807           14   
7929  23/12/2022 00:00     12808           20   
7930  23/12/2022 00:00     12809           14   
7931  23/12/2022 00:00     12810           14   

                       descripcion_producto  cantidad  precio_producto  \
0     LICUADORA MAN 7 VELOCIDADES PLASTICO.         1           1500.0   
1                       FRAZADA PROVIPOLAR.         1            400.0   
2           JUEGO DE SaBANAS MATRIMONIALES.         1            430.0   
3                 colcha thalia matrimonial         1            380.0   
4                COBERTOR EXCEL MATRIMONIAL         1            800.0   
...                                     ...       ...              ...   
7927             COBERTOR JUMBO MATRIMONIAL         1           1400.0   
7928             COBERTOR JUMBO MATRIMONIAL         1           1400.0   
7929          COORDINADO PERCAL MATRIMONIAL         1           1850.0   
7930             COBERTOR JUMBO MATRIMONIAL         2           1400.0   
7931             COBERTOR JUMBO MATRIMONIAL         1           1400.0   

      importe  monto_total  monto_abono condicion_pago  ... id_pueblo  \
0         0.0        975.0            0        Semanal  ...        29   
1       400.0        800.0            0        Semanal  ...        29   
2         0.0        450.0            0        Semanal  ...        29   
3         0.0        929.0            0        Semanal  ...        60   
4         0.0        480.0            0        Semanal  ...        62   
...       ...          ...          ...            ...  ...       ...   
7927      0.0       1400.0           50        Semanal  ...        21   
7928      0.0       1400.0           50        Semanal  ...         3   
7929      0.0       3250.0          100        Semanal  ...         3   
7930      0.0       2800.0          100        Semanal  ...         3   
7931      0.0       1400.0           50        Semanal  ...         3   

       poblacion estatus_cliente                       lugar_trabajo  \
0        ATLIXCO         Premium                    VENTA DE MOLOTES   
1        ATLIXCO          Moroso                         TORTOLLERIA   
2        ATLIXCO         Regular                         AMA DE CASA   
3     EL PARAISO         Regular  VENTA DE TAMALES ENTRADA A SABANA.   
4         TEYUCA          Moroso                        AMA DE CASA.   
...          ...             ...                                 ...   
7927         NaN         Regular                         ama de casa   
7928         NaN         Regular                         ama de casa   
7929         NaN         Regular                         ama de casa   
7930         NaN         Regular                         ama de casa   
7931         NaN         Regular                            empleado   

     id_vendedor  nombre_vendedor apaterno_vendedor estatus_vendedor  \
0              1          ANTONIO         GUTIERREZ             Baja   
1              3          GERARDO               NaN             Baja   
2              3          GERARDO               NaN             Baja   
3             12            FELIX          MARTINEZ           Activo   
4              2            DAVID           COTZOMI             Baja   
...          ...              ...               ...              ...   
7927          23            PABLO         VELAZQUEZ           Activo   
7928          23            PABLO         VELAZQUEZ           Activo   
7929          23            PABLO         VELAZQUEZ           Activo   
7930          12            FELIX          MARTINEZ           Activo   

# Observación de los Datos

Al observar los datos se muestra que tenemos 124658 filas y 28 columnas. Es una base de datos de ventas en la cual tenemos columnas importantes como la fecha de la venta, el producto con su respectivo id o identificador, la descripción del producto, el precio del producto.

Usando Phyton  con sus librerias pandas y numpy y analizando los datos podríamos sacar diferentes datos estadísticos pero primero hay que limpiar los datos.

Con la siguiente funcion ***dtypes*** vamos a revisar el nombre y el tipo de cada columna y eliminar algunas que no nos interesen.

In [ ]:
df.dtypes

fecha_venta              object
id_venta                  int64
id_producto               int64
descripcion_producto     object
cantidad                  int64
precio_producto         float64
importe                 float64
monto_total             float64
monto_abono               int64
condicion_pago           object
estatus_venta            object
id_cliente                int64
nombre_cliente           object
apaterno_cliente         object
colonia                  object
id_pueblo                 int64
poblacion                object
estatus_cliente          object
lugar_trabajo            object
id_vendedor               int64
nombre_vendedor          object
apaterno_vendedor        object
estatus_vendedor         object
domicilio                object
observaciones            object
dtype: object

#Limpieza de Datos

Al observar los datos, nos percatamos de que hay columnas que nos dan poca información o que no son de mucho interéscomo por ejemplo monto_abono, condicion_pago, domicilio y observaciones. Las cuales vamos a eliminar en el siguiente bloque de código utilizando la funcion **drop()**.

In [ ]:
nuevo_df= df.drop(columns = ['monto_abono','condicion_pago','colonia','domicilio','observaciones','observaciones']) 
nuevo_df

fecha_venta  id_venta  id_producto  \
0     24/10/2013 00:00         1           38   
1     10/08/2013 00:00         2           28   
2     24/12/2012 00:00         3           34   
3     14/10/2013 00:00         6           15   
4     15/10/2013 00:00         7           12   
...                ...       ...          ...   
7927  23/12/2022 00:00     12806           14   
7928  23/12/2022 00:00     12807           14   
7929  23/12/2022 00:00     12808           20   
7930  23/12/2022 00:00     12809           14   
7931  23/12/2022 00:00     12810           14   

                       descripcion_producto  cantidad  precio_producto  \
0     LICUADORA MAN 7 VELOCIDADES PLASTICO.         1           1500.0   
1                       FRAZADA PROVIPOLAR.         1            400.0   
2           JUEGO DE SaBANAS MATRIMONIALES.         1            430.0   
3                 colcha thalia matrimonial         1            380.0   
4                COBERTOR EXCEL MATRIMONIAL         1            800.0   
...                                     ...       ...              ...   
7927             COBERTOR JUMBO MATRIMONIAL         1           1400.0   
7928             COBERTOR JUMBO MATRIMONIAL         1           1400.0   
7929          COORDINADO PERCAL MATRIMONIAL         1           1850.0   
7930             COBERTOR JUMBO MATRIMONIAL         2           1400.0   
7931             COBERTOR JUMBO MATRIMONIAL         1           1400.0   

      importe  monto_total estatus_venta  id_cliente nombre_cliente  \
0         0.0        975.0        Pagada           1       HIPOLITA   
1       400.0        800.0       Deudora           2  NANCY LIZBETH   
2         0.0        450.0        Pagada           3      FELICITAS   
3         0.0        929.0        Pagada           6        MODESTA   
4         0.0        480.0        Pagada           7          REYNA   
...       ...          ...           ...         ...            ...   
7927      0.0       1400.0       Deudora        6860        eloina    
7928      0.0       1400.0       Deudora        6861    rosa isela    
7929      0.0       3250.0       Deudora        6862       maricela   
7930      0.0       2800.0       Deudora        6863      leonarda    
7931      0.0       1400.0       Deudora        6864          cesar   

     apaterno_cliente  id_pueblo   poblacion estatus_cliente  \
0            MARTINEZ         29     ATLIXCO         Premium   
1             NEGRETE         29     ATLIXCO          Moroso   
2             CAMACHO         29     ATLIXCO         Regular   
3             DELGADO         60  EL PARAISO         Regular   
4            GONZALEZ         62      TEYUCA          Moroso   
...               ...        ...         ...             ...   
7927         gonzalez         21         NaN         Regular   
7928        gutierrez          3         NaN         Regular   
7929         alvarez           3         NaN         Regular   
7930          alvarez          3         NaN         Regular   
7931          mayorga          3         NaN         Regular   

                           lugar_trabajo  id_vendedor nombre_vendedor  \
0                       VENTA DE MOLOTES            1         ANTONIO   
1                            TORTOLLERIA            3         GERARDO   
2                            AMA DE CASA            3         GERARDO   
3     VENTA DE TAMALES ENTRADA A SABANA.           12           FELIX   
4                           AMA DE CASA.            2           DAVID   
...                                  ...          ...             ...   
7927                         ama de casa           23           PABLO   
7928                         ama de casa           23           PABLO   
7929                         ama de casa           23           PABLO   
7930                         ama de casa           12           FELIX   
7931                            empleado           23           PABLO   

     apaterno_vendedor estatus_ven

Vamos a eliminar los datos nulos para limpieza de los mismos.

In [ ]:
nuevo_df.dropna()

fecha_venta  id_venta  id_producto  \
0     24/10/2013 00:00         1           38   
3     14/10/2013 00:00         6           15   
4     15/10/2013 00:00         7           12   
5     15/10/2013 00:00        11           39   
6     28/10/2013 00:00        16           69   
...                ...       ...          ...   
7920  21/12/2022 00:00     12799           94   
7921  21/12/2022 00:00     12800           94   
7923  21/12/2022 00:00     12802           61   
7925  20/12/2022 00:00     12804           94   
7926  20/12/2022 00:00     12805          230   

                              descripcion_producto  cantidad  precio_producto  \
0            LICUADORA MAN 7 VELOCIDADES PLASTICO.         1           1500.0   
3                        colcha thalia matrimonial         1            380.0   
4                       COBERTOR EXCEL MATRIMONIAL         1            800.0   
5     LICUADORA OSTERIZER 10 VELOCIDADES PLASTICO.         1           1200.0   
6                       COLCHA FLORAL MATRIMONIAL.         1            450.0   
...                                            ...       ...              ...   
7920                      COBERTOR  AParizzio MAT.         1           1400.0   
7921                      COBERTOR  AParizzio MAT.         2           1400.0   
7923        CAMA tactopiel C/COLCHON holliday MAT.         1           9100.0   
7925                      COBERTOR  AParizzio MAT.         1           1400.0   
7926                     CUBRECOLCHON SOGNARE MAT.         1           2700.0   

      importe  monto_total estatus_venta  id_cliente   nombre_cliente  \
0         0.0        975.0        Pagada           1         HIPOLITA   
3         0.0        929.0        Pagada           6          MODESTA   
4         0.0        480.0        Pagada           7            REYNA   
5         0.0        950.0        Pagada          11          BRIGIDA   
6       315.0        315.0        Pagada          16        FRANCISCA   
...       ...          ...           ...         ...              ...   
7920      0.0       1400.0       Deudora        6845           gisela   
7921      0.0       2800.0       Deudora        2556  MARIA DEL ROCIO   
7923      0.0       9100.0       Deudora        4819    maria cecilia   
7925      0.0       1400.0       Deudora        6170           dayami   
7926      0.0       2700.0       Deudora        6370            irene   

     apaterno_cliente  id_pueblo      poblacion estatus_cliente  \
0            MARTINEZ         29        ATLIXCO         Premium   
3             DELGADO         60     EL PARAISO         Regular   
4            GONZALEZ         62         TEYUCA          Moroso   
5             CHARROS         62         TEYUCA         Regular   
6           COMISARIO         61        HUILUCO         Regular   
...               ...        ...            ...             ...   
7920             alva         72     tepapayeca         Regular   
7921           TORRES         81  CACALOXUCHITL         Regular   
7923          vilchis         88       vallarta         Regular   
7925           torres         83       TEACALCO         Regular   
7926           cortez         87        bonilla         Regular   

                           lugar_trabajo  id_vendedor nombre_vendedor  \
0                       VENTA DE MOLOTES            1         ANTONIO   
3     VENTA DE TAMALES ENTRADA A SABANA.           12           FELIX   
4                           AMA DE CASA.            2           DAVID   
5                           AMA DE CASA.            1         ANTONIO   
6                           AMA DE CASA.           12           FELIX   
...                                  ...          ...             ...   
7920                         ama de casa           12           FELIX   
7921                        AMA DE CASA.           12           FELIX   
7923                         ama de casa           12           FELIX   
7925                         ama de c

#Transformación de Datos

Quiero manipular el campo fecha_venta pero actualmente es de tipo Objet. Para poderlo usar lo cambiaremos de tipo usando la libreria **DateTime**.

In [ ]:
import datetime
from datetime import datetime #importamos la libreria datetime

nuevo_df['fecha_venta'] = pd.to_datetime(nuevo_df['fecha_venta'], errors='coerce')
nuevo_df.dtypes

fecha_venta             datetime64[ns]
id_venta                         int64
id_producto                      int64
descripcion_producto            object
cantidad                         int64
precio_producto                float64
importe                        float64
monto_total                    float64
estatus_venta                   object
id_cliente                       int64
nombre_cliente                  object
apaterno_cliente                object
id_pueblo                        int64
poblacion                       object
estatus_cliente                 object
lugar_trabajo                   object
id_vendedor                      int64
nombre_vendedor                 object
apaterno_vendedor               object
estatus_vendedor                object
dtype: object

Ahora fecha_venta ya es un campo tipo datetime.

In [ ]:
nuevo_df['Mes'] = nuevo_df['fecha_venta'].dt.month
nuevo_df['Año'] = nuevo_df['fecha_venta'].dt.year
nuevo_df.head(2) # El año y Mes aparecen en la última columna.

fecha_venta  id_venta  id_producto                   descripcion_producto  \
0  2013-10-24         1           38  LICUADORA MAN 7 VELOCIDADES PLASTICO.   
1  2013-10-08         2           28                    FRAZADA PROVIPOLAR.   

   cantidad  precio_producto  importe  monto_total estatus_venta  id_cliente  \
0         1           1500.0      0.0        975.0        Pagada           1   
1         1            400.0    400.0        800.0       Deudora           2   

   ... id_pueblo poblacion  estatus_cliente     lugar_trabajo id_vendedor  \
0  ...        29   ATLIXCO          Premium  VENTA DE MOLOTES           1   
1  ...        29   ATLIXCO           Moroso       TORTOLLERIA           3   

  nombre_vendedor  apaterno_vendedor estatus_vendedor   Mes     Año  
0         ANTONIO          GUTIERREZ             Baja  10.0  2013.0  
1         GERARDO                NaN             Baja  10.0  2013.0  

[2 rows x 22 columns]

Como podemos observar, en el dataframe anterior se agregaron dos nuevas columas utilizando las funciones *dt.month* y *dt.year* de la libreria **DateTime**.

He querido manipular el campo estatus_venta, pero al parecer no se ha podido porque es de tipo Object, en el siguiente bloque se realizará el cambio del tipo de dato a **String**.

In [ ]:
nuevo_df['estatus_venta'] = nuevo_df['estatus_venta'].astype('string')
nuevo_df.dtypes

fecha_venta             datetime64[ns]
id_venta                         int64
id_producto                      int64
descripcion_producto            object
cantidad                         int64
precio_producto                float64
importe                        float64
monto_total                    float64
estatus_venta                   string
id_cliente                       int64
nombre_cliente                  object
apaterno_cliente                object
id_pueblo                        int64
poblacion                       object
estatus_cliente                 object
lugar_trabajo                   object
id_vendedor                      int64
nombre_vendedor                 object
apaterno_vendedor               object
estatus_vendedor                object
Mes                            float64
Año                            float64
dtype: object

# Análisis de Datos

Una vez que los datos se limpiaron y transformaron podemos empezar con un pequeño análisis. Obteniendo la moda de las columnas se puede obtener cierta información.

In [ ]:
nuevo_df['id_producto'].mode()  
nuevo_df['descripcion_producto'].mode()
nuevo_df['estatus_cliente'].mode()
nuevo_df['lugar_trabajo'].mode()
nuevo_df['nombre_vendedor'].mode()
nuevo_df['Mes'].mode()
nuevo_df['Año'].mode()

0    2022.0
dtype: float64

Del código anterior podemos concluir lo siguiente:

*   El producto que más se vendió es el que tiene el id **14**.
*   El producto que más se ha vendido es el **Cobertor Jumbo Matrimonial**.
*   La mayoria de los clientes tienen un estatus **Regular**.
*   La ocupación mas frecuente del cliente es ser **Ama de casa**.
*   El vendedor que más vendio se llama **Felix**
*   El mes en el que mas se vende es **Junio**.
*   El año que más se ha vendido es en el **2022**.

De los datos anteriores podemos hacer un análisis del año 2022 que fue en el que mas se vendió. Por lo tanto quisiera saber cuanto dinero se obtuvo en el año 2022 con es estatus de las ventas igual a Pagado.

In [ ]:
nuevo_df.loc[(nuevo_df['estatus_venta'] == 'Pagada') & (nuevo_df['Año'] == 2022), 'monto_total'].sum()

483776.0

HRMuebles recibió la cantidad de **$483,776.00**

Tambien quiero saber cuantos deudores estaban registrados en el mes de enero del 2021.

In [ ]:
nuevo_df.loc[(nuevo_df['estatus_venta'] == 'Deudora') & (nuevo_df['Mes'] == 1) & (nuevo_df['Año'] == 2021), 'monto_total'].sum()

29850.0

HRMuebles tuvo **29850** deudores en enero del 2021

#Gráficos

Visualización de datos de las ventas que realizaron los trabajadores en el año 2022 y el producto que se vendió.

In [ ]:
año = nuevo_df[nuevo_df['Año']==2022]

In [ ]:
import altair as alt
import matplotlib.pyplot as plt

alt.Chart(año).mark_bar().encode(
    y=alt.Y("nombre_vendedor", title='Vendedor'),
    x=alt.X("descripcion_producto", title='Producto Vendido'),
    color=alt.Color("Mes", title="Mes del Año", scale=alt.Scale(scheme='reds')),
).properties(
    title='Ventas de productos realizadas por los vendedores en el 2022'
)



alt.Chart(...)

Se observa que solo hubo 7 vendedores activos en el 2022. 

In [ ]:

alt.Chart(año).mark_circle().encode(
    y=alt.Y("poblacion", title='Población'),
    x=alt.X("descripcion_producto", title='Producto Vendido'),
    color=alt.Color("Mes", title="Mes del Año", scale=alt.Scale(scheme='reds')),
).properties(
    title='Productos vendidos en las localidades del Sur de Puebla en el Año 2022'
)

alt.Chart(...)

Del gráfico anterior podemos concluir que en el 2022 se vendieron más productos en las localidades de Atlixco, Huaquechula, Cacaloxuchitl, Tepeojuma y Teacalco. Y que uno de los productos que más se vende en el suroeste del Estado de Puebla es la licuadora y el Cobertor Jumbo.